### PR TITLE
get and show api character

### PR DESCRIPTION
**Pull request by phamvanbk11**

### CODE REVIEW

File: app/src/main/java/com/fstyle/androidtrainning/data/source/remote/MarvelRemote.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package com.fstyle.androidtrainning.data.source.remote;
+
+import retrofit2.Retrofit;
+import retrofit2.converter.gson.GsonConverterFactory;
+
+/**
+ * Created by framgia on 19/12/2017.
+ */
+
+public class MarvelRemote {
```
Comment (trainer-android-fr) [inline app/src/main/java/com/fstyle/androidtrainning/data/source/remote/MarvelRemote.java]: WARNING: Utility classes should not have a public or default constructor.

File: app/src/main/java/com/fstyle/androidtrainning/data/source/remote/api/service/APIService.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package com.fstyle.androidtrainning.data.source.remote.api.service;
+
+
+import com.fstyle.androidtrainning.data.source.remote.api.response.MarvelResponse;
+
+import retrofit2.Call;
+import retrofit2.http.GET;
+
+public interface APIService {
+    @GET("/v1/public/characters?apikey=7a4ee917fb53941ccc691445dac51ef7&hash=d68af7429d2507d9bd9e21af20b412f2&ts=1513072715795")
```
Comment (trainer-android-fr) [inline app/src/main/java/com/fstyle/androidtrainning/data/source/remote/api/service/APIService.java]: ERROR: Line is longer than 100 characters (found 128).

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/MainActivity.java [hunk 3]
```diff
@@ -2,33 +2,52 @@
...
+        showAllCharacter();
+    }
+
+    private void showAllCharacter() {
+        APIService apiService = MainApplication.getMarvel().create(APIService.class);
+        Call<MarvelResponse> call = apiService.getRespone();
+        call.enqueue(new Callback<MarvelResponse>() {
+            @Override
+            public void onResponse(Call<MarvelResponse> call, Response<MarvelResponse> response) {
+                List<CharacterItem> result_character = response.body().getData().getResults();
```
Comment (trainer-android-fr) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/MainActivity.java]: WARNING: Name 'result_character' must match pattern '^[a-z][a-zA-Z0-9]*$'.

File: app/src/main/java/com/fstyle/androidtrainning/data/source/remote/MarvelRemote.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package com.fstyle.androidtrainning.data.source.remote;
+
+import retrofit2.Retrofit;
+import retrofit2.converter.gson.GsonConverterFactory;
+
+/**
+ * Created by framgia on 19/12/2017.
+ */
+
+public final class MarvelRemote {
```
Comment (trainer-android-fr) [inline app/src/main/java/com/fstyle/androidtrainning/data/source/remote/MarvelRemote.java]: WARNING: Utility classes should not have a public or default constructor.

File: app/src/main/java/com/fstyle/androidtrainning/data/source/remote/api/service/APIService.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package com.fstyle.androidtrainning.data.source.remote.api.service;
+
+
+import com.fstyle.androidtrainning.data.source.remote.api.response.MarvelResponse;
+
+import retrofit2.Call;
+import retrofit2.http.GET;
+
+public interface APIService {
+    @GET("/v1/public/characters?apikey=7a4ee917fb53941ccc691445dac51ef7&" +
```
Comment (trainer-android-fr) [inline app/src/main/java/com/fstyle/androidtrainning/data/source/remote/api/service/APIService.java]: WARNING: '+' should be on a new line.

File: app/src/main/java/com/fstyle/androidtrainning/data/source/remote/api/service/APIService.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+import com.fstyle.androidtrainning.data.source.remote.api.response.MarvelResponse;
+
+import retrofit2.Call;
+import retrofit2.http.GET;
+
+public interface APIService {
+    @GET("/v1/public/characters?apikey=7a4ee917fb53941ccc691445dac51ef7&"
+            + "hash=d68af7429d2507d9bd9e21af20b412f2&ts=1513072715795")
+    Call<MarvelResponse> getRespone();
+}
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/data/source/remote/api/service/APIService.java]: add 1 blank line when end of file

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/ItemViewModel.java [hunk 7]
```diff
@@ -24,6 +28,18 @@ public String getName() {
     public String getDescription() {
         return mCharacter.getDescription();
     }
+    public String getImageUrl() {
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/ItemViewModel.java]: and 1 blank line above this line!

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/ItemViewModel.java [hunk 8]
```diff
@@ -24,6 +28,18 @@ public String getName() {
...
+        return mCharacter.getThumbnail().getPath() + "." + mCharacter.getThumbnail().getExtension();
+    }
+
+    @BindingAdapter({"app:imageUrl"})
+    public static void loadImage(ImageView view, String imageUrl) {
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/ItemViewModel.java]: move this method into file have name ``BindingUtils``

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/MainActivity.java [hunk 9]
```diff
@@ -2,33 +2,52 @@
...
-        mViewModel = new MainCharacterViewModel(mainAdapter);
-        CharacterMainBinding binding =
-                DataBindingUtil.setContentView(this, R.layout.character_main);
-        binding.setViewModel(mViewModel);
+        showAllCharacter();
+    }
+
+    private void showAllCharacter() {
+        APIService apiService = MainApplication.getMarvel().create(APIService.class);
+        Call<MarvelResponse> call = apiService.getRespone();
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/MainActivity.java]: apiService.getRespone();
-> 
Pls rename method ``getRespone`` to getListHeroes``

File: app/src/main/res/layout/item_character.xml [hunk 10]
```diff
@@ -25,24 +26,25 @@
 
                 <ImageView
                     android:id="@+id/character_image"
-                    android:layout_width="wrap_content"
-                    android:layout_height="wrap_content"
+                    android:layout_width="48sp"
```
Comment (daolq3012) [inline app/src/main/res/layout/item_character.xml]: hard code!
48sp -> 48dp, sp only use for the size of text
pls define in dimens.xml and use it in here!

File: app/src/main/res/layout/item_character.xml [hunk 11]
```diff
@@ -25,24 +26,25 @@
...
+                    android:layout_height="48sp"
                     android:layout_alignParentStart="true"
                     android:layout_alignParentLeft="true"
                     android:layout_alignParentTop="true"
                     android:layout_marginRight="16dp"
```
Comment (daolq3012) [inline app/src/main/res/layout/item_character.xml]: hard code

File: app/src/main/res/layout/item_character.xml [hunk 12]
```diff
@@ -25,24 +26,25 @@
...
                     android:layout_width="wrap_content"
                     android:layout_height="wrap_content"
                     android:layout_alignParentTop="true"
-                    android:layout_toRightOf="@+id/character_image"
                     android:layout_toEndOf="@+id/character_image"
-                    android:textSize="30sp"
-                    android:text="@{viewModel.name}"/>
+                    android:layout_toRightOf="@+id/character_image"
+                    android:text="@{viewModel.name}"
+                    android:textSize="30sp" />
```
Comment (daolq3012) [inline app/src/main/res/layout/item_character.xml]: hard code

File: app/src/main/res/layout/item_character.xml [hunk 13]
```diff
@@ -25,33 +26,37 @@
 
                 <ImageView
                     android:id="@+id/character_image"
-                    android:layout_width="wrap_content"
-                    android:layout_height="wrap_content"
-                    android:layout_alignParentStart="true"
+                    android:layout_width="48dp"
```
Comment (daolq3012) [inline app/src/main/res/layout/item_character.xml]: .

File: app/src/main/res/layout/item_character.xml [hunk 14]
```diff
@@ -25,33 +26,37 @@
...
+                    android:contentDescription="@string/content_description_image"
+                    app:imageUrl="@{viewModel.imageUrl}" />
 
                 <TextView
                     android:id="@+id/character_name"
                     android:layout_width="wrap_content"
                     android:layout_height="wrap_content"
                     android:layout_alignParentTop="true"
-                    android:layout_toRightOf="@+id/character_image"
+                    android:layout_marginLeft="13dp"
```
Comment (daolq3012) [inline app/src/main/res/layout/item_character.xml]: .

File: app/src/main/res/layout/item_character.xml [hunk 15]
```diff
@@ -25,33 +26,37 @@
...
+                    android:textSize="?attr/dialogPreferredPadding" />
 
                 <TextView
                     android:id="@+id/character_description"
                     android:layout_width="wrap_content"
                     android:layout_height="wrap_content"
-                    android:layout_below="@+id/character_name"
-                    android:layout_toRightOf="@+id/character_image"
+                    android:layout_below="@+id/character_image"
+                    android:layout_marginLeft="14dp"
```
Comment (daolq3012) [inline app/src/main/res/layout/item_character.xml]: .

File: app/src/main/res/values/dimens.xml [hunk 16]
```diff
@@ -0,0 +1,10 @@
+<resources>
+    <!-- Item Character -->
+    <dimen name="image_width">48dp</dimen>
+    <dimen name="image_height">48dp</dimen>
+    <dimen name="name_margin_left">13dp</dimen>
+    <dimen name="name_margin_start">13dp</dimen>
+    <dimen name="description_margin_left">14dp</dimen>
```
Comment (trainer-android-fr) [inline app/src/main/res/values/dimens.xml]: WARNING: [UnusedResources] The resource `R.dimen.description_margin_left` appears to be unused
 Unused resources make applications larger and slow down builds.

File: app/src/main/res/values/dimens.xml [hunk 17]
```diff
@@ -0,0 +1,10 @@
...
+    <dimen name="description_margin_start">14dp</dimen>
```
Comment (trainer-android-fr) [inline app/src/main/res/values/dimens.xml]: WARNING: [UnusedResources] The resource `R.dimen.description_margin_start` appears to be unused
 Unused resources make applications larger and slow down builds.

File: app/src/main/res/values/dimens.xml [hunk 18]
```diff
@@ -0,0 +1,10 @@
...
+    <dimen name="description_margin_top">10dp</dimen>
```
Comment (trainer-android-fr) [inline app/src/main/res/values/dimens.xml]: WARNING: [UnusedResources] The resource `R.dimen.description_margin_top` appears to be unused
 Unused resources make applications larger and slow down builds.

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/MainActivity.java [hunk 19]
```diff
@@ -2,33 +2,52 @@
...
-        MainAdapter mainAdapter = new MainAdapter(this, (characters));
-        mViewModel = new MainCharacterViewModel(mainAdapter);
-        CharacterMainBinding binding =
-                DataBindingUtil.setContentView(this, R.layout.character_main);
-        binding.setViewModel(mViewModel);
+        showAllCharacter();
+    }
+
+    private void showAllCharacter() {
+        APIService apiService = MainApplication.getMarvel().create(APIService.class);
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/MainActivity.java]: Pls inject APIService into MainCharacterViewModel and call inside it!

**Pull request merged**